### PR TITLE
Handle internal server errors from charmhub

### PR DESCRIPTION
**Pull request by SimonRichardson**
Server errors from charmhub can return back plain/text mime types, even
though we tell the server we only accept json. To better improve the
error message we check the status code and return back if it's a 5xx
error.

The user never has to see the dump from the server and we can get the
dump from the response in the server logs.

We also ensure that we close the body and allow the response to be
reused after draining.

----

As a driveby, we introduce a MIME type alias because it's painful to see
"application/json" everywhere.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju find k8s
$ juju find
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1924195

### CODE REVIEW

File: charmhub/http.go [hunk 1]
```diff
@@ -83,9 +91,19 @@ func (t *APIRequester) Do(req *http.Request) (*http.Response, error) {
 	var potentialInvalidURL bool
 	if resp.StatusCode == http.StatusNotFound {
 		potentialInvalidURL = true
+	} else if resp.StatusCode >= http.StatusInternalServerError && resp.StatusCode <= http.StatusNetworkAuthenticationRequired {
+		defer func() {
+			_, _ = io.Copy(ioutil.Discard, resp.Body)
+			_ = resp.Body.Close()
+		}()
+		return nil, errors.Errorf(`unexpected server error %q`, req.URL.String())
```
Comment (manadart) [inline charmhub/http.go:99]: I don't think we need "unexpected" here. Should we include the status code we got?

**Review by manadart — APPROVED** [review]
Minor suggestion.

File: charmhub/http.go (hunk 1 above)
Comment (SimonRichardson) [reply charmhub/http.go:99]: We dump the response further up, so I don't want to expose internal details.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**